server/src/main/kotlin/org/javacs/kt/overridemembers/OverrideMembers.kt

```diff
@@ -0,0 +1,282 @@
+                        when (member) {
+                            is FunctionDescriptor -> createFunctionStub(member)
+                            is PropertyDescriptor -> createVariableStub(member)
+                            else -> null
+                        }
+                    }
+            } else {
+                null
+            }
+        }
+        .flatten()
+
+private fun ClassDescriptor.canBeExtended() = this.kind.isInterface ||
+    this.modality == Modality.ABSTRACT ||
+    this.modality == Modality.OPEN
+
+private fun FunctionDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
+
+private fun PropertyDescriptor.canBeOverriden() = (Modality.ABSTRACT == this.modality || Modality.OPEN == this.modality) && Modality.FINAL != this.modality && this.visibility != DescriptorVisibilities.PRIVATE && this.visibility != DescriptorVisibilities.PROTECTED
+            
+// interfaces are ClassDescriptors by default. When calling AbstractClass super methods, we get a ClassConstructorDescriptor
+fun getClassDescriptor(descriptor: DeclarationDescriptor?): ClassDescriptor? =
+        if (descriptor is ClassDescriptor) {
+            descriptor
+        } else if (descriptor is ClassConstructorDescriptor) {
+            descriptor.containingDeclaration
+        } else {
+            null
+        }
+
+fun getSuperClassTypeProjections(
+        file: CompiledFile,
+        superType: KtSuperTypeListEntry
+): List<TypeProjection> =
+        superType
+                .typeReference
+                ?.typeElement
+                ?.children
+                ?.filter { it is KtTypeArgumentList }
+                ?.flatMap { (it as KtTypeArgumentList).arguments }
+                ?.mapNotNull {
+                    (file.referenceExpressionAtPoint(it?.startOffset ?: 0)?.second as?
+                                    ClassDescriptor)
+                            ?.defaultType?.asTypeProjection()
+                }
+                ?: emptyList()
+
+// Checks if the class overrides the given declaration
+fun overridesDeclaration(kotlinClass: KtClass, descriptor: FunctionDescriptor): Boolean =
+        kotlinClass.declarations.any {
+            if (it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
+            ) {
+                if (it is KtNamedFunction) {
+                    parametersMatch(it, descriptor)
+                } else {
+                    true
+                }
+            } else {
+                false
+            }
+        }
+
+fun overridesDeclaration(kotlinClass: KtClass, descriptor: PropertyDescriptor): Boolean =
+        kotlinClass.declarations.any {
+            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
+        }
```

Comment:
If I am not mistaken, we can shorten this to

```suggestion
// Checks if the class overrides the given declaration
fun overridesDeclaration(kotlinClass: KtClass, descriptor: FunctionDescriptor): Boolean =
        kotlinClass.declarations.any {
            it.name == descriptor.name.asString()
            && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
            && ((it as? KtNamedFunction)?.let { parametersMatch(it, descriptor) } ?? true)
        }

fun overridesDeclaration(kotlinClass: KtClass, descriptor: PropertyDescriptor): Boolean =
        kotlinClass.declarations.any {
            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
        }
```

Maybe we could even merge these overloads:

```suggestion
// Checks if the class overrides the given declaration
fun overridesDeclaration(kotlinClass: KtClass, descriptor: MemberDescriptor): Boolean =
    when (descriptor) {
        is FunctionDescriptor -> kotlinClass.declarations.any {
            it.name == descriptor.name.asString()
            && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
            && ((it as? KtNamedFunction)?.let { parametersMatch(it, descriptor) } ?? true)
        }
        is PropertyDescriptor -> kotlinClass.declarations.any {
            it.name == descriptor.name.asString() && it.hasModifier(KtTokens.OVERRIDE_KEYWORD)
        }
        else -> false
    }
```

That would get simplify the `.filter { classMember -> ... }` above at the expense of a bit of type-safety, so I am not sure whether that would actually be more elegant.

Reply:
It prettifies a lot by making the code more compact, so I did that. See my comment above on the `filter`-expression and what the input is. We would probably have to smart cast to MemberDescriptor for the simplification to work. Unless I'm missing something. Will take another look tomorrow morning, as I'm starting to get tired now.

Reply:
I was missing something 🙂 Took another look after getting some sleep, and found that just casting to `MemberDescriptor` in the filter-expression makes the expression more compact and expressive.  Done in [596c6a3](https://github.com/fwcd/kotlin-language-server/pull/379/commits/596c6a3268e2fa6d9cbeb13daa551e03af464e54).